### PR TITLE
Wip/hughsie/config defaults

### PR DESCRIPTION
**Pull request by hughsie**
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

**Comment by hughsie** [comment]
@superm1 another idea, that nukes the defines completely.

**Comment by superm1** [comment]
Yeah I'm happy with this direction.

BTW - It's of course a bit risky, but I think we should bring all our config file handling fixes and maybe even the "empty" conffile stuff back to 1_9_X.  Realistically 1.9.13 is probably what's going to be in Ubuntu 24.04 and RHEL 9.5 and such so we really want to get right to avoid bugs for the next 2 years about "my upgrade did XYZ".

**Comment by hughsie** [comment]
I was thinking of lifting your self tests on top, then I'd feel a lot more confident about back porting.

**Comment by superm1** [comment]
> I was thinking of lifting your self tests on top, then I'd feel a lot more confident about back porting.

Yeah I'll rebase them after this PR is done.  We can talk about what the backport looks like after that.

**Comment by superm1** [comment]
The difficult part about the self tests as it pertains to this PR is that the defaults are set in plugins and the engine both now, but the unit self tests nominally wouldn't load all the plugins and engine.  I _guess_ migration tests need to be an installed test not a unit test unless you've got some other ideas here.

**Comment by hughsie** [comment]
> unless you've got some other ideas here.

Fake news! I just hardcoded the plugin defaults in the self tests.

**Comment by superm1** [comment]
I mean the full conf migration tests I had in the other PR

**Comment by hughsie** [comment]
> I mean the full conf migration tests I had in the other PR

I think I added those, no?

File: src/fu-self-test.c [hunk 1]
```diff
@@ -6010,6 +6028,145 @@ fu_remote_local_func(void)
...
+	g_assert_no_error(error);
+	g_assert_true(ret);
+	g_assert_cmpstr(localconf_data,
+			==,
+			"[fwupd]\n"
+			"AllowEmulation=true\n");
+}
+
+static void
+fu_config_set_plugin_defaults(FuConfig *config)
```
Comment (superm1) [inline src/fu-self-test.c:6080]: I guess this works but it's unfortunate we still have everything duplicated both in code and tests.

How about a private function definition we can call for all the defaults in self tests and in code instead.

**Comment by superm1** [comment]
> > I mean the full conf migration tests I had in the other PR
> 
> I think I added those, no?

I must have looked at older version before and missed them now. Yes, looks mostly good. Just one comment.

File: src/fu-self-test.c (hunk 1 above)
Comment (hughsie) [reply src/fu-self-test.c:6080]: > I guess this works but it's unfortunate we still have everything duplicated

So maybe it's overkill to migrate all the config files? How about just the MSR one? And then it's just one line of duplication.

Comment (superm1) [reply src/fu-self-test.c:6080]: I go back and forth. How about if we record the old md5sums of unmodified and just remove unmodified ones?

Comment (hughsie) [reply src/fu-self-test.c:6080]: Oh no, what I mean is maybe we just migrate msr.conf in the tests.

Comment (superm1) [reply src/fu-self-test.c:6080]: Oh like just one?  Would say migrate redfish if just one. That's the one that sploded.m in reports.

Comment (hughsie) [reply src/fu-self-test.c:6080]: Okay, I was thinking about this a bit more -- and I do think we want to duplicate the values -- as it actually makes sense to decouple the "what the daemon default should be" and "what the self test expects the value should be" -- for instance, if we change the MSR `MinimumSmeKernelVersion` from `5.18.0` to `5.19.0` in the plugin we want the old default to be ignored too.

**Review by superm1 — APPROVED** [review]

**Pull request merged**